### PR TITLE
SMR: API support for Dedicated Servers

### PR DESCRIPTION
**Pull request by porisius**
Support for Dedicated Servers to upload, scan, and download for mods and SML.

Special thanks to Vilsol, Feyko, Night, Mircea, Peter, and others for the encouragement and constant questions.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -86,5 +86,5 @@ func initializeDefaults() {
 
 	viper.SetDefault("frontend.url", "")
 
-	viper.SetDefault("virustotal.key", "")
+	viper.SetDefault("virustotal.key", "739c8578a77d9d34c07f77cb92cdc4835236d3162698ae6eabfd5fccad6b89d7")
```
Comment (Vilsol) [inline config/config.go:89]: Please remove the virustotal key from defaults

File: validation/validation.go [hunk 2]
```diff
@@ -346,6 +346,11 @@ func validateUPluginJSON(archive *zip.Reader, uPluginFile *zip.File, withValidat
 					Path: file.Name,
 					Type: "sml_mod",
 				})
+			} else if extension == "so" {
+				modInfo.Objects = append(modInfo.Objects, ModObject{
+					Path: file.Name,
+					Type: "sml_mod",
+				})
```
Comment (Vilsol) [inline validation/validation.go:353]: Merge this into the if statement above with `extension == "dll" || extension == "so"`

File: schemas/mod_links.graphql [hunk 3]
```diff
@@ -0,0 +1,56 @@
+### Types
```
Comment (Vilsol) [inline schemas/mod_links.graphql:1]: The file should probably be renamed to `mod_archs.graphql` and all objects should follow the same convention as per the rename.

File: schemas/sml_links.graphql [hunk 4]
```diff
@@ -0,0 +1,60 @@
+### Types
```
Comment (Vilsol) [inline schemas/sml_links.graphql:1]: The file should probably be renamed to `sml_archs.graphql` and all objects should follow the same convention as per the rename.

File: redis/jobs/consumers/consumer_scan_mod_on_virus_total.go [hunk 5]
```diff
@@ -40,7 +40,12 @@ func ScanModOnVirusTotalConsumer(ctx context.Context, payload []byte) error {
 	log.Ctx(ctx).Info().Msgf("starting virus scan of mod %s version %s", task.ModID, task.VersionID)
 
 	version := postgres.GetVersion(ctx, task.VersionID)
-	link := storage.GenerateDownloadLink(version.Key)
+	mod := postgres.GetModByID(ctx, version.ModID)
+
+	//link := storage.GenerateDownloadLink(version.Key)
```
Comment (Vilsol) [inline redis/jobs/consumers/consumer_scan_mod_on_virus_total.go:45]: Please remove the unused code

File: storage/storage.go [hunk 6]
```diff
@@ -289,6 +293,49 @@ func DeleteMod(ctx context.Context, modID string, name string, versionID string)
...
+		if err := storage.Delete(key); err != nil {
+			log.Ctx(ctx).Err(err).Msg("failed to delete version")
+			return false
+		}
+	}
+
+	return true
+}
+
+func DeleteModLink(ctx context.Context, modID string, name string, versionID string, platform string) bool {
```
Comment (Vilsol) [inline storage/storage.go:323]: Should be renamed to `DeleteModArch`

File: storage/storage.go [hunk 7]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+func SeparateMod(ctx context.Context, body []byte, modID, name string, versionID string, modVersion string) (bool, string) {
+	//read combined file
+	zipReader, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
+
+	if err != nil {
+		return false, ""
+	}
+
+	// Create a buffer to write our archive to.
+	log.Ctx(ctx).Info().Msg("Beginning Splitting Operations")
```
Comment (Vilsol) [inline storage/storage.go:408]: Please add more context to the log e.g. the mod id and version

File: storage/storage.go [hunk 8]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to write file to zip: %w", err)
+	}
+
+	return nil
+}
+
+func WriteModLink(ctx context.Context, key string, versionID string, platform string, buffer *bytes.Buffer) error {
```
Comment (Vilsol) [inline storage/storage.go:567]: Rename to `WriteModArch`

File: storage/storage.go [hunk 9]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
 	}
 	return result
 }
+
+func SeparateMod(ctx context.Context, body []byte, modID, name string, versionID string, modVersion string) (bool, string) {
```
Comment (Vilsol) [inline storage/storage.go:399]: Please convert this into a function that loops over an array consisting of `LinuxServer`, `WindowsServer` and `WindowsNoEditor`, so you don't duplicate the same code 3x.

File: storage/storage.go [hunk 10]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		Platform:         platform,
+		Key:              key,
+		Hash:             hex.EncodeToString(hash.Sum(nil)),
+		Size:             int64(len(buffer.Bytes())),
+	}
+
+	_, err = postgres.CreateModLink(ctx, dbModLink)
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:594]: Please remove the debug println as you return an error with context

File: storage/storage.go [hunk 11]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		return fmt.Errorf("failed to write smod: %w", err)
+	}
+
+	hash := sha256.New()
+	_, err = hash.Write(buffer.Bytes())
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:579]: Please remove the debug println as you return an error with context

File: storage/storage.go [hunk 12]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+	}
+
+	return nil
+}
+
+func WriteModLink(ctx context.Context, key string, versionID string, platform string, buffer *bytes.Buffer) error {
+	_, err := storage.Put(ctx, key, bytes.NewReader(buffer.Bytes()))
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:571]: Please remove the debug println as you return an error with context

File: storage/storage.go [hunk 13]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		return fmt.Errorf("failed to read zipped file: %w", err)
+	}
+
+	_, err = zipFile.Write(buf.Bytes())
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:560]: Please use the logging library

File: storage/storage.go [hunk 14]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		return fmt.Errorf("failed to open zipped file: %w", err)
+	}
+
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(rawFile)
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:553]: Please remove the debug println as you return an error with context

File: storage/storage.go [hunk 15]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		return fmt.Errorf("failed to create zip: %w", err)
+	}
+
+	rawFile, err := file.Open()
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:545]: Please remove the debug println as you return an error with context

File: storage/storage.go [hunk 16]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+
+	return true, key
+}
+
+func WriteZipFile(file *zip.File, platform string, zipWriter *zip.Writer) error {
+	fileName := strings.ReplaceAll(file.Name, platform+"/", "")
+	zipFile, err := zipWriter.Create(fileName)
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:538]: Please remove the debug println as you return an error with context

File: storage/storage.go [hunk 17]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+
+	key := fmt.Sprintf("/mods/%s/%s.smod", modID, cleanName+"-Combined-"+modVersion)
+	platform := "Combined"
+
+	combined := bytes.NewBuffer(body)
+
+	err = WriteModLink(ctx, key, versionID, platform, combined)
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:524]: Please log the error using the logging library

File: storage/storage.go [hunk 18]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		}
+	}
+	if Win64Client {
+		key := fmt.Sprintf("/mods/%s/%s.smod", modID, cleanName+"-WindowsNoEditor-"+modVersion)
+		platform := "WindowsNoEditor"
+
+		err = WriteModLink(ctx, key, versionID, platform, bufWin64Client)
+
+		if err != nil {
+			fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:511]: Please log the error using the logging library

File: storage/storage.go [hunk 19]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+	}
+
+	if Win64Server {
+		key := fmt.Sprintf("/mods/%s/%s.smod", modID, cleanName+"-Win64Server-"+modVersion)
+		platform := "Win64Server"
+
+		err = WriteModLink(ctx, key, versionID, platform, bufWin64Server)
+
+		if err != nil {
+			fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:500]: Please log the error using the logging library

File: storage/storage.go [hunk 20]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+	}
+
+	//Write to mod_link and upload new smaller smod file
+	if LinuxServer {
+		key := fmt.Sprintf("/mods/%s/%s.smod", modID, cleanName+"-LinuxServer-"+modVersion)
+		platform := "LinuxServer"
+
+		err = WriteModLink(ctx, key, versionID, platform, bufLinuxServer)
+		if err != nil {
+			fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:488]: Please log the error using the logging library

File: storage/storage.go [hunk 21]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		return false, ""
+	}
+	err = zipWriterWin64Server.Close()
+	if err != nil {
+		fmt.Println(err)
+		return false, ""
+	}
+	err = zipWriterWin64Client.Close()
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:477]: Please log the error using the logging library

File: storage/storage.go [hunk 22]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+
+	// Make sure to check the error on Close.
+	err = zipWriterLinuxServer.Close()
+	if err != nil {
+		fmt.Println(err)
+		return false, ""
+	}
+	err = zipWriterWin64Server.Close()
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:472]: Please log the error using the logging library

File: storage/storage.go [hunk 23]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+			}
+
+			Win64Client = true
+		}
+	}
+
+	// Make sure to check the error on Close.
+	err = zipWriterLinuxServer.Close()
+	if err != nil {
+		fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:467]: Please log the error using the logging library

File: storage/storage.go [hunk 24]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+			}
+
+			Win64Server = true
+		}
+
+		if strings.Contains(file.Name, "WindowsNoEditor") {
+			err = WriteZipFile(file, "WindowsNoEditor", zipWriterWin64Client)
+
+			if err != nil {
+				fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:456]: Please log the error using the logging library

File: storage/storage.go [hunk 25]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+			}
+
+			LinuxServer = true
+		}
+
+		if strings.Contains(file.Name, "WindowsServer") {
+			err = WriteZipFile(file, "WindowsServer", zipWriterWin64Server)
+
+			if err != nil {
+				fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:445]: Please log the error using the logging library

File: storage/storage.go [hunk 26]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+	for _, file := range zipReader.File {
+		if strings.HasPrefix(file.Name, ".pdb") || strings.HasPrefix(file.Name, ".debug") || strings.Contains(file.Name, modVersion) {
+			continue
+		}
+
+		if strings.Contains(file.Name, "LinuxServer") {
+			err = WriteZipFile(file, "LinuxServer", zipWriterLinuxServer)
+
+			if err != nil {
+				fmt.Println(err)
```
Comment (Vilsol) [inline storage/storage.go:434]: Please log the error using the logging library

File: storage/storage.go [hunk 27]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		Key:              key,
+		Hash:             hex.EncodeToString(hash.Sum(nil)),
+		Size:             int64(len(buffer.Bytes())),
+	}
+
+	_, err = postgres.CreateModLink(ctx, dbModLink)
+
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to create ModLink: %w", err)
```
Comment (Vilsol) [inline storage/storage.go:595]: Please wrap errors using `errors.Wrap(err, "...")`

File: storage/storage.go [hunk 28]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		fmt.Println(err)
+		return fmt.Errorf("failed to write smod: %w", err)
+	}
+
+	hash := sha256.New()
+	_, err = hash.Write(buffer.Bytes())
+
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to hash smod: %w", err)
```
Comment (Vilsol) [inline storage/storage.go:580]: Please wrap errors using `errors.Wrap(err, "...")`

File: storage/storage.go [hunk 29]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+
+	return nil
+}
+
+func WriteModLink(ctx context.Context, key string, versionID string, platform string, buffer *bytes.Buffer) error {
+	_, err := storage.Put(ctx, key, bytes.NewReader(buffer.Bytes()))
+
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to write smod: %w", err)
```
Comment (Vilsol) [inline storage/storage.go:572]: Please wrap errors using `errors.Wrap(err, "...")`

File: storage/storage.go [hunk 30]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to read zipped file: %w", err)
+	}
+
+	_, err = zipFile.Write(buf.Bytes())
+
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to write file to zip: %w", err)
```
Comment (Vilsol) [inline storage/storage.go:561]: Please wrap errors using `errors.Wrap(err, "...")`

File: storage/storage.go [hunk 31]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+		fmt.Println(err)
+		return fmt.Errorf("failed to open zipped file: %w", err)
+	}
+
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(rawFile)
+
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to read zipped file: %w", err)
```
Comment (Vilsol) [inline storage/storage.go:554]: Please wrap errors using `errors.Wrap(err, "...")`

File: storage/storage.go [hunk 32]
```diff
@@ -348,3 +395,205 @@ func EncodeName(name string) string {
...
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to create zip: %w", err)
+	}
+
+	rawFile, err := file.Open()
+
+	if err != nil {
+		fmt.Println(err)
+		return fmt.Errorf("failed to open zipped file: %w", err)
```
Comment (Vilsol) [inline storage/storage.go:546]: Please wrap errors using `errors.Wrap(err, "...")`

File: db/postgres/mod_links.go [hunk 33]
```diff
@@ -0,0 +1,121 @@
+package postgres
```
Comment (Vilsol) [inline db/postgres/mod_links.go:1]: Please rename the file to `mod_archs.go` and all the appropriate names as well.

File: db/postgres/postgres_types.go [hunk 34]
```diff
@@ -202,3 +204,19 @@ type Compatibility struct {
 	State string
 	Note  string
 }
+
+type ModLink struct {
```
Comment (Vilsol) [inline db/postgres/postgres_types.go:208]: Please rename to `ModArch`

File: db/postgres/postgres_types.go [hunk 35]
```diff
@@ -202,3 +204,19 @@ type Compatibility struct {
...
+	ID               string `gorm:"primary_key;type:varchar(16)"`
+	ModVersionLinkID string
+	Platform         string
+	Key              string
+	Size             int64
+	Hash             string
+}
+
+type SMLLink struct {
```
Comment (Vilsol) [inline db/postgres/postgres_types.go:217]: Please rename to `SMLArch`

File: db/postgres/sml_links.go [hunk 36]
```diff
@@ -0,0 +1,90 @@
+package postgres
```
Comment (Vilsol) [inline db/postgres/sml_links.go:1]: Please rename the file to `sml_archs.go` and all the appropriate names as well.

File: gql/resolver_mod_links.go [hunk 37]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func (r *queryResolver) GetModLinks(ctx context.Context, filter map[string]interface{}) (*generated.GetModLinks, error) {
+	wrapper, _ := WrapQueryTrace(ctx, "getModLinks")
+	defer wrapper.end()
+	return &generated.GetModLinks{}, nil
+}
+
+func (r *queryResolver) GetModLinkByID(ctx context.Context, modLinkID string) (*generated.ModLink, error) {
+	wrapper, newCtx := WrapQueryTrace(ctx, "getModLink")
```
Comment (Vilsol) [inline gql/resolver_mod_links.go:76]: Change the trace name to `getModLinkByID`

File: gql/versions.go [hunk 38]
```diff
@@ -124,9 +124,12 @@ func FinalizeVersionUploadAsync(ctx context.Context, mod *postgres.Mod, versionI
 	}
 
 	// TODO Validate mod files
-	success, key := storage.RenameVersion(ctx, mod.ID, mod.Name, versionID, modInfo.Version)
+	//Okay, uploaded file is read and readable... let's dump it and separate, and re-save as ModName-Combined-SemVersion in ModLink
+	storage.DeleteVersion(ctx, mod.ID, mod.Name, versionID)
 
-	if !success {
+	separated, key := storage.SeparateMod(ctx, fileData, mod.ID, mod.Name, dbVersion.ID, modInfo.Version)
```
Comment (Vilsol) [inline gql/versions.go:130]: Is this correct? You are deleting the version and then splitting it?

File: migrations/sql/000021_add_mod_platform.up.sql [hunk 39]
```diff
@@ -0,0 +1,21 @@
+-- Mod Links
+create table if not exists mod_links
```
Comment (Vilsol) [inline migrations/sql/000021_add_mod_platform.up.sql:2]: Please rename to `mod_archs`

File: migrations/sql/000021_add_mod_platform.up.sql [hunk 40]
```diff
@@ -0,0 +1,21 @@
...
+    mod_version_link_id     varchar(14),
+    platform                varchar(16),
+    key                     text,
+    hash                    char(64),
+    size                    bigint
+);
+create index if not exists idx_mod_links_id on mod_links (mod_version_link_id, platform);
+
+-- SML Links
+create table if not exists sml_links
```
Comment (Vilsol) [inline migrations/sql/000021_add_mod_platform.up.sql:14]: Please rename to `sml_archs`

File: schemas/mod_links.graphql [hunk 41]
```diff
@@ -0,0 +1,56 @@
...
+
+    getModDownload(ModVersionLinkID: String!, platform: String!): String!
+}
+
+### Mutations
+
+extend type Mutation {
+    createModLink(mod_link: NewModLink!): ModLink @isLoggedIn
+    deleteModLink(ModLinksId: ModLinkID!): Boolean! @canEditMod(field: "versionId") @isLoggedIn
+}
```
Comment (Vilsol) [inline schemas/mod_links.graphql:56]: It should not be possible to modify version files after initial upload.

**Review by Vilsol — CHANGES_REQUESTED** [review]

File: gql/versions.go [hunk 42]
```diff
@@ -124,9 +124,12 @@ func FinalizeVersionUploadAsync(ctx context.Context, mod *postgres.Mod, versionI
 	}
 
 	// TODO Validate mod files
```
Comment (budak7273) [inline gql/versions.go:126]: Should this TODO still be here?

File: nodes/mod.go [hunk 43]
```diff
@@ -288,3 +288,33 @@ func downloadModVersion(c echo.Context) error {
...
+
+	if version == nil {
+		return c.String(404, "version not found")
+	}
+
+	platform := postgres.GetModArchByPlatform(c.Request().Context(), versionID, platformType)
+
+	if platform == nil {
+		return c.String(404, "platform not found")
+	}
```
Comment (budak7273) [inline nodes/mod.go:313]: Would there be any benefit to including the modId, versionId, and platformType in the error messages that are returned?

File: nodes/mod.go [hunk 44]
```diff
@@ -288,3 +288,33 @@ func downloadModVersion(c echo.Context) error {
 
 	return c.Redirect(302, storage.GenerateDownloadLink(version.Key))
 }
+
+func downloadModVersionArch(c echo.Context) error {
```
Comment (budak7273) [inline nodes/mod.go:292]: This function appears to lack a doc comment, whereas all others in this file have one

File: nodes/version.go [hunk 45]
```diff
@@ -52,3 +52,26 @@ func downloadVersion(c echo.Context) error {
 
 	return c.Redirect(302, storage.GenerateDownloadLink(version.Key))
 }
+
+func downloadModArch(c echo.Context) error {
```
Comment (budak7273) [inline nodes/version.go:56]: Same doc comment + error message comments I left on `nodes/mod.go`

File: gql/versions.go (hunk 42 above)
Comment (Vilsol) [reply gql/versions.go:126]: This is in current codebase as well, it can be removed.

Comment (porisius) [reply gql/versions.go:126]: I'll remove it then in a commit.

File: validation/validation.go (hunk 2 above)
Comment (porisius) [reply validation/validation.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/c92ca90e3a9e65269f3fa50c8ff3d9637e571008

File: storage/storage.go (hunk 7 above)
Comment (porisius) [reply storage/storage.go]: Removed as it was used for when I was first splitting so I had a log of where I was...
https://github.com/satisfactorymodding/smr-api/pull/12/commits/0938ddafb99e9ba28c8d21ee0c6a45f34d7be867

File: storage/storage.go (hunk 10 above)
Comment (porisius) [reply storage/storage.go]: Heads up on lint, I had to leave fmt.Errorf in due to lint... the error was...
storage/storage.go:533:10: error returned from external package is unwrapped: sig: func (*archive/zip.File).Open() (io.ReadCloser, error) (wrapcheck)
                return err
https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 11 above)
Comment (porisius) [reply storage/storage.go]: Heads up on lint, I had to leave fmt.Errorf in due to lint... the error was...
storage/storage.go:533:10: error returned from external package is unwrapped: sig: func (*archive/zip.File).Open() (io.ReadCloser, error) (wrapcheck)
                return err
https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 12 above)
Comment (porisius) [reply storage/storage.go]: Heads up on lint, I had to leave fmt.Errorf in due to lint... the error was...
storage/storage.go:533:10: error returned from external package is unwrapped: sig: func (*archive/zip.File).Open() (io.ReadCloser, error) (wrapcheck)
                return err
https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 13 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 14 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 15 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 16 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 17 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 18 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 19 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 20 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 21 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 22 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 24 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 23 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 25 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: storage/storage.go (hunk 26 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/74cbf9e7fb628b477e7bb2459a0f641867ec9415

File: gql/versions.go (hunk 42 above)
Comment (porisius) [reply gql/versions.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/0dcaad74bf3b745b8367792fb6fe0fc989a6ba23

File: schemas/mod_links.graphql (hunk 41 above)
Comment (porisius) [reply schemas/mod_links.graphql]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/8ea4f2b30882e41af00e95ce37ab41b3a4ebd4e3

File: migrations/sql/000021_add_mod_platform.up.sql (hunk 40 above)
Comment (porisius) [reply migrations/sql/000021_add_mod_platform.up.sql]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/8ea4f2b30882e41af00e95ce37ab41b3a4ebd4e3

File: migrations/sql/000021_add_mod_platform.up.sql (hunk 39 above)
Comment (porisius) [reply migrations/sql/000021_add_mod_platform.up.sql]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/8ea4f2b30882e41af00e95ce37ab41b3a4ebd4e3
https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: gql/versions.go (hunk 38 above)
Comment (porisius) [reply gql/versions.go]: Deleting, splitting, then saving it again with a ModArch to send over for validation...

The delete and rename, essentially, is no different from storage.RenameVersion. I can reuse that code if you desire.

File: nodes/mod.go (hunk 43 above)
Comment (porisius) [reply nodes/mod.go:313]: I don't see a problem with it, but bear in mind, I did copy from downloadModVersion. I have the change ready if you desire me to commit/push it with both of them changed.

File: gql/resolver_mod_links.go (hunk 37 above)
Comment (porisius) [reply gql/resolver_mod_links.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: schemas/sml_links.graphql (hunk 4 above)
Comment (porisius) [reply schemas/sml_links.graphql]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: schemas/mod_links.graphql (hunk 3 above)
Comment (porisius) [reply schemas/mod_links.graphql]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: storage/storage.go (hunk 6 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: db/postgres/mod_links.go (hunk 33 above)
Comment (porisius) [reply db/postgres/mod_links.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: db/postgres/postgres_types.go (hunk 34 above)
Comment (porisius) [reply db/postgres/postgres_types.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: db/postgres/sml_links.go (hunk 36 above)
Comment (porisius) [reply db/postgres/sml_links.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: db/postgres/postgres_types.go (hunk 35 above)
Comment (porisius) [reply db/postgres/postgres_types.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: storage/storage.go (hunk 32 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/38c4e9639463832b010c55fa96c058828ec8210e

File: storage/storage.go (hunk 31 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/38c4e9639463832b010c55fa96c058828ec8210e

File: storage/storage.go (hunk 30 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/38c4e9639463832b010c55fa96c058828ec8210e

File: storage/storage.go (hunk 29 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/38c4e9639463832b010c55fa96c058828ec8210e

File: storage/storage.go (hunk 28 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/38c4e9639463832b010c55fa96c058828ec8210e

File: storage/storage.go (hunk 27 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/38c4e9639463832b010c55fa96c058828ec8210e

File: nodes/version.go (hunk 45 above)
Comment (porisius) [reply nodes/version.go:66]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/5e5146206a7fe78acf4f87620417a535e6a9a579

File: nodes/mod.go (hunk 44 above)
Comment (porisius) [reply nodes/mod.go:302]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/5e5146206a7fe78acf4f87620417a535e6a9a579

File: nodes/mod.go (hunk 43 above)
Comment (porisius) [reply nodes/mod.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/5e5146206a7fe78acf4f87620417a535e6a9a579

File: storage/storage.go (hunk 8 above)
Comment (porisius) [reply storage/storage.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/bdd565d8997568955b8e33bcb352c38c7af5c2c5

File: redis/jobs/consumers/consumer_scan_mod_on_virus_total.go (hunk 5 above)
Comment (porisius) [reply redis/jobs/consumers/consumer_scan_mod_on_virus_total.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/8ea4f2b30882e41af00e95ce37ab41b3a4ebd4e3

File: gql/versions.go (hunk 38 above)
Comment (porisius) [reply gql/versions.go]: Also, at that point in the code, the file has been read into memory, I am just sending the buffer/bytes to SeparateMod that I am going to revisit to cut down repeat code.

Comment (Vilsol) [reply gql/versions.go]: In the `RenameVersion` function it calls to the underlying `Rename` which actually just copies the object, and THEN it deletes it from storage, not the other way around.

Comment (porisius) [reply gql/versions.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/2ac1243aed7bd5d4641f015af08a919157fc46a0
https://github.com/satisfactorymodding/smr-api/pull/12/commits/c078aa8776d72806e5fb249eef0a7d77d31c3929

Let me know if this is up to par... ignore cleanPlatform, it was a foolish idea, and I have removed it.

File: storage/storage.go (hunk 9 above)
Comment (porisius) [reply storage/storage.go:399]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/2ac1243aed7bd5d4641f015af08a919157fc46a0
https://github.com/satisfactorymodding/smr-api/pull/12/commits/c078aa8776d72806e5fb249eef0a7d77d31c3929

Let me know if this is up to par... ignore cleanPlatform, it was a foolish idea, and I have removed it.

File: gql/versions.go (hunk 38 above)
Comment (porisius) [reply gql/versions.go]: https://github.com/satisfactorymodding/smr-api/pull/12/commits/dd24599d4a6af63705883b7b89dc47b956820c32

Let me know if this is more desired.

**Review by Vilsol — APPROVED** [review]
LGTM for staging

**Pull request merged**